src/util/triplea/MapXMLCreator/CanalDefinitionsPanel.java

```diff
@@ -0,0 +1,442 @@
+            // continue;
+            LinkedHashSet<String> landTerrNeighbors2 = MapXMLHelper.territoryConnections.get(landTerr2);
+            if (landTerrNeighbors2 == null)
+              landTerrNeighbors2 = new LinkedHashSet<String>();
+            // if (landTerrNeighbors2.contains(landTerr))
+            // continue;
+            if (!landTerrNeighbors.contains(landTerr2) && !landTerrNeighbors2.contains(landTerr))
+              continue;
+            final SortedSet<String> waterTerrs2 = new TreeSet<String>(landWaterTerrConn2.getValue());
+            waterTerrs2.retainAll(waterTerrs);
+            if (waterTerrs2.size() > 1) {
+              // remove water territories that are not connected to the other water territories
+              final SortedSet<String> waterTerrs2Copy = new TreeSet<String>(waterTerrs2);
+              for (Iterator<String> iter_waterTerr2 = waterTerrs2.iterator(); iter_waterTerr2.hasNext();) {
+                final String waterTerr2 = (String) iter_waterTerr2.next();
+                waterTerrs2Copy.remove(waterTerr2);
+                final SortedSet<String> waterTerrs2ReqNeighbors = new TreeSet<String>(waterTerrs2Copy);
+                final LinkedHashSet<String> waterTerr2Neightbors = MapXMLHelper.territoryConnections.get(waterTerr2);
+                if (waterTerr2Neightbors != null) {
+                  for (final String waterTerr2Neighbor : waterTerr2Neightbors)
+                    waterTerrs2ReqNeighbors.remove(waterTerr2Neighbor);
+                }
+                if (!waterTerrs2ReqNeighbors.isEmpty())
+                  iter_waterTerr2.remove();
+              }
+              // create canal only if at least 2 water territories remain
+              if (waterTerrs2.size() > 1) {
+                final SortedSet<String> newLandSet = new TreeSet<String>();
+                newLandSet.add(landTerr);
+                newLandSet.add(landTerr2);
+                final Tuple<SortedSet<String>, SortedSet<String>> terrTuple =
+                    Tuple.of(new TreeSet<String>(waterTerrs2), newLandSet);
+                MapXMLHelper.putCanalDefinitions("Canal" + MapXMLHelper.canalDefinitions.size(), terrTuple);
+              }
+            }
+          }
+        }
+        boolean noNewCanalsBuild = MapXMLHelper.canalDefinitions.isEmpty();
+        if (noNewCanalsBuild)
+          JOptionPane.showMessageDialog(null, "No canals have been build!", "Auto-Fill Result",
+              JOptionPane.PLAIN_MESSAGE);
+        else {
+          final StringBuilder sb = new StringBuilder();
+          sb.append("<html>The following " + MapXMLHelper.canalDefinitions.size() + " canals have been build:");
+          for (final Entry<String, Tuple<SortedSet<String>, SortedSet<String>>> canalDef : MapXMLHelper.canalDefinitions
+              .entrySet()) {
+            sb.append("<br/> - " + canalDef.getKey() + ": ");
+            Iterator<String> iter_waterTerr = canalDef.getValue().getFirst().iterator();

```

Comment:
Should be an apache common `StringUtils.join(canalDef.getValue(), "-")` that you cana use. Guava has similar with the `Joiner` api